### PR TITLE
Generate code from java kafka client

### PR DESCRIPTION
**Pull request by codesnik**

### CODE REVIEW

**Comment by codesnik** [comment]
This is update for https://github.com/klarna/kafka_protocol/pull/21
It's a work in progress right now, I've yet to use it once with running kafka, and I've never touched brod yet.

But it already compiles, yay!

**Comment by codesnik** [comment]
So, biggest changes are in generated code, and it's `.gitignore`d.
I'm thinking, codegen like this better be checked into git, just to be able to better see what's really changed between versions. Still, here's the gist :)

it was like this,
https://gist.github.com/codesnik/1626c26dd2d9cae68923721c8c271152/e9063a44b8a32782ca3d2549e40fbe67c4c6ef43

```erlang
-record(kpro_FetchRequest,
        { replicaId :: kpro:int32()
        , maxWaitTime :: kpro:int32()
        , minBytes :: kpro:int32()
        , fetchRequestTopic_L :: [kpro_FetchRequestTopic()]
        }).

-record(kpro_FetchRequestTopic,
        { topicName :: kpro:str()
        , fetchRequestPartition_L :: [kpro_FetchRequestPartition()]
        }).

-record(kpro_FetchRequestPartition,
        { partition :: kpro:int32()
        , fetchOffset :: kpro:int64()
        , maxBytes :: kpro:int32()
        }).
```

now it's like this:
https://gist.github.com/codesnik/1626c26dd2d9cae68923721c8c271152

```erlang
-record(kpro_fetch_request_v0,
        { replica_id :: kpro:int32()
        , max_wait_time :: kpro:int32()
        , min_bytes :: kpro:int32()
        , topics :: [kpro_fetch_request_v0_topic()]
        }).

-record(kpro_fetch_request_v0_topic,
        { topic :: kpro:str()
        , partitions :: [kpro_fetch_request_v0_partition()]
        }).

-record(kpro_fetch_request_v0_partition,
        { partition :: kpro:int32()
        , fetch_offset :: kpro:int64()
        , max_bytes :: kpro:int32()
        }).
```

**Comment by zmstone** [comment]
very impressive work!
one quick comment before I look into the details:
according to elvis coding style (https://github.com/inaka/elvis)
it's better to define `-type some_type() :: #some_record{}` in a `.erl` file and have them exported like `-export_type([some_type/0])`.
Now it's the perfect time to make this change.

I am for putting generated code to repo, let's do that.

**Comment by codesnik** [comment]
I've probably messed up all the streaming logic, I can only hope brod tests would allow me to catch this.

@zmstone I'm not sure I understand how to apply export_type-ing suggestion to kafka_protocol. First, that erl file still should be generated, so it's probably won't be a src/kpro.erl, right? And all the types would go to some other module? How to name it then?
Should record definitions still go into include/kpro.hrl? our types aren't opaque, records still should be usable from brod somehow. Maybe I'm missing something?

**Comment by codesnik** [comment]
@zmstone also, I'd like to hear your opinion about naming scheme. Just only because structure of subrecords _might_ change, I have to version every subrecord. (I'm already happy that field names are unique across all the subrecords of particular record, so I don't have to namespace everything more than once, this could be awful).

But this leads to otherwise identical `#kpro_fetch_request_v0_partition{}`, `#kpro_fetch_request_v1_partition` and so on. Right now I updated code in src/kpro.erl to use record versions it already used before, no version magic. But I don't know how to add that magic without copypasting every record constructor for every supported version.

**Comment by zmstone** [comment]
record definition should be put to `include/krpo.hrl`
only the type specs should be moved to `kpro.erl`
for example the last type spec in the generated `kpro.hrl` is
```
-type kpro_delete_topics_response_v0_topic_error_code() ::
#kpro_delete_topics_response_v0_topic_error_code{}.
```
elvis recommends in `kpro.erl`:
```
-export_type([delete_topics_response_v0/0]).
...
...
...
-type delete_topics_response_v0_topic_error_code() ::
#kpro_delete_topics_response_v0_topic_error_code{}.
```

**Comment by zmstone** [comment]
ah, since kpro.erl is hand-coded, we'll have to put the generated specs to another module kpro_types.erl
I'm not sure now.
let's drop this for now, I have to think about it more.

**Comment by codesnik** [comment]
@zmstone problem is, kpro.erl isn't generated (yet?) :) only kpro.hrl is. And I really don't like the idea of making kpro.hrl generated, at least with tools I already know (maybe rebar3 has some kind of usable file templates?). Also types are messed up anyway, I don't think dializer or other tools like that would catch all the {incomplete, ...} and {already_encoded...} magic.

**Comment by zmstone** [comment]
there is no template for erlang code generation, (not that i know of :) ) you can generate AST and have it pretty printed, but that's more magic than it is now.
I hate having to manually craft the indentations and everything too, if you see other tools e.g. Python fits better, feel free to use it.
You don't even have to use the yrl and xrl magic if that's sort of too much magic too.

**Comment by codesnik** [comment]
I don't need template for code generation, just simple generic text template would work, mustache etc. But looks like there's nothing like that in standard library.

I already spent some time fiddling with leex and yecc, and I think I already invested too much to drop it now :) If you look into diff, you'll see I made parser even more magical! And I have an idea or two to make it even mooore magical.

python generate erlang code using bnf generated by java code! That'd be convoluted.

One question though: can I try to port this to rebar3? Would this be a problem for a project using rebar2, depending on brod and kafka_protocol? I'm not sure how dependencies building works in rebar2/rebar3 yet.

**Comment by zmstone** [comment]
rebar3 is sort of supported already, but go ahead move code generation to to rebar hooks
(if that's what you mean)
However, we need to be backward compatible for rebar2 users (using rebar.config.script)

**Comment by zmstone** [comment]
sorry, missed this one:

> Just only because structure of subrecords might change, I have to version every subrecord. (I'm already happy that field names are unique across all the subrecords of particular record, so I don't have to namespace everything more than once, this could be awful).

You mean like `kpro_fetch_response_v0_partition_header` and `kpro_fetch_response_v1_partition_header` ?
yes, it would be nice to have only `kpro_fetch_response_partition_header` if all fields are the same.

maybe we should use a different scheme for sub-records? e.g. prop-list ? or maps ?

**Comment by codesnik** [comment]
@zmstone I DON'T KNOW! :) brod and kafka_protocol are the only erlang projects I've ever seen, I don't know what works and what not, what's hot and what's dated :)

Fields could be the same in versions 0-3 and add just one field in some upcoming version 4, and that's it, it won't longer work. But records definitely saved me a lot of typos when I changed the code by compile time checks, and fields codegen works properly with them. One solution I proposed earlier in #21 was to make "merged records", which would drop versioning from record definitions, making them interchangeable, but carrying around some "undefined"'s. This would even retain some compile time checks.

**Comment by zmstone** [comment]
I don't know either, I haven't seen any code generation projects that are forced to deal this many schema/protocol versions.

Merge fields is perhaps better than maps or prop-lists, let's try that
but please continue doing what you think is the best way forward.

I'll also play with the new code a bit to have a better understanding on the pros and cons of different approaches.

File: include/kpro_common.hrl [hunk 1]
```diff
@@ -45,118 +45,107 @@
...
 %% correlation IDs are 32 bit signed integers.
-%% we use 27 bits only, and use the highest 5 bits to be redudant with API key
+%% we use 25 bits only, and use the highest 5 bits to be redudant with API key
+%% and next 2 bits with API version
 %% so that the decoder may decode the responses without the need of an extra
 %% correlation ID to API key association.
--define(CORR_ID_BITS, 27).
--define(MAX_CORR_ID, 134217727).
+-define(API_KEY_BITS, 5).
+-define(API_VERSION_BITS, 2).
```
Comment (zmstone) [inline include/kpro_common.hrl:53]: make it 3.

**Comment by zmstone** [comment]
@id @codesnik 
I have pushed a quick PoC to #29 
generated code looks like this:
https://gist.github.com/zmstone/9d446b5e8caab92ef91ac135aec439f5

Public APIs may look like
```
kpro:encode(MessageName, VersionNumber, DataEitherProplistOrMap) ->
  encode(kpro_schema:get(MessageName, VersionNumber), DataEitherProplistOrMap)).
```
```
kpro:decode(CorrId, Binary) ->
  {MessageName, Version} = decode_corr_id(CorrId),
  Schema = kpro_schema:get(MessageName, Version),
  decode(Schema, Binary).
```
It has its drawbacks, but at least there won't be a need for copy-pasting 
the long `foo_vX_bar_and_baz` record names when constructing the records.

File: src/kpro.erl [hunk 2]
```diff
@@ -188,26 +190,32 @@ decode_response(Bin, Acc) ->
...
 
-%% @doc help function to encode kpro_XxxRequest into kafka wire format.
--spec encode_request(client_id(), corr_id(), kpro_RequestMessage()) -> iodata().
+%% @doc help function to encode kpro_*_request_v* into kafka wire format.
+-spec encode_request(client_id(), corr_id(), kpro_request_message()) -> iodata().
 encode_request(ClientId, CorrId, Request) ->
-  R = #kpro_Request{ correlationId  = CorrId
-                   , clientId       = ClientId
-                   , requestMessage = Request
+  %% FIXME huston, we have a problem
```
Comment (id) [inline src/kpro.erl:196]: houston

File: src/kpro.erl [hunk 3]
```diff
@@ -357,50 +359,61 @@ encode({{array, T}, L}) when is_list(L) ->
 encode({array, L}) when is_list(L) ->
   Length = length(L),
   [<<Length:32/?INT>>, [encode(I) || I <- L]];
-encode(#kpro_PartitionMessageSet{} = R) ->
+%% FIXME encoder could be better? also what about  ...v1_data and ...v2_data
+%% probably should be done with encode(records, Bin) for RECORDS type)
+encode(#kpro_produce_request_v0_data{} = R) ->
   %% messages in messageset is a stream, not an array
-  EncodedMessages = R#kpro_PartitionMessageSet.message_L,
+  %% FIXME what is the difference between stream and array here?
```
Comment (id) [inline src/kpro.erl:366]: Arrays in kafka protocol are preceded by int32 containing number of elements.
In fetch response message sets are preceded by int32 containing byte size of the set.

File: README.md [hunk 4]
```diff
@@ -1,10 +1,17 @@
...
+This is a kafka wire format encode/decode library, not a kafka client.
+See https://github.com/klarna/brod for kafka client implementation.
+
+Code generated from org.apache.kafka.common.protocol.Protocol.
+
+## How to generate kafka.bnf
+Ensure you have java (1.7+) and gradle (2.0+) installed.
+Change kafka version in priv/kafka_protocol_bnf/build.gradle if needed.
 
+  (cd priv/kafka_protocol_bnf && gradle -q run) > priv/kafka.bnf
```
Comment (id) [inline README.md:12]: add this to Makefile

Comment (codesnik) [reply README.md:12]: that was my intent. Though, I'd rather not run this for every build, only when file's missing.

File: src/kpro.erl (hunk 3 above)
Comment (codesnik) [reply src/kpro.erl:366]: ah, now I see. thanks.

File: README.md (hunk 4 above)
Comment (id) [reply README.md:12]: sure, just make `priv/kafka.bnf` a dependency for compiling

**Comment by zmstone** [comment]
included in #21 @codesnik thanks for your awesome work.

**Comment by codesnik** [comment]
@zmstone thanks for actually using it!

**Pull request closed without merging**